### PR TITLE
Fix interpolate -limit Add interpolate limit_direction='inside'

### LINKED ISSUE #16282: df.interpolate limit_direction= does not work without limit=x setting
the interpolate limit_direction setting should work even without a limit=x setting.

When interpolate is called without a limit=x setting, it always (apparently) uses the limit_direction setting default of 'forward' even when it is explicitly set to 'backward' or 'both'

The code below will show 3 DataFrames.

- The first shows the initial before interpolation.
- The second shows the interpolation done correctly in the backward direction.
- The third shows the dataframe done incorrectly in the 'forward' direction even though the limit_direction='backward' was used. 

The docs do not mention a linkage between the two settings, but if this is expected behavior (I hope not), it would be oh so nice to raise an exception like the existing error for a bad limit_direction setting.

Many thanks for all your work!

#### Example Code:

```python
import pandas as pd
import numpy as np

dfMain = pd.DataFrame({
    'a': [0, 1, np.NAN, 3, 4],
    'b': [np.NaN, 1, np.NaN, 3, 4],
    'c': [0 , 1, 2, 3, np.NaN]})

print("initial df")
print(dfMain)

print("backwards with limit=1 (correct result):")
print(dfMain.interpolate(limit_direction='backward', limit=1))

print("backwards no longer works when limit is removed (it is going forward):")
print(dfMain.interpolate(limit_direction='backward'))
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-75-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 34.4.1
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.1
matplotlib: 2.0.0
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.2
bs4: 4.5.3
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.5
pymysql: None
psycopg2: None
jinja2: 2.9.4
boto: 2.45.0
pandas_datareader: 0.2.1
None
initial df

</details>

### PR DESCRIPTION
**Pull request by WBare**
- [ x] closes #16282 and #16284
 - [x ] tests added / passed
 - [ x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -20,6 +20,12 @@ Check the :ref:`API Changes <whatsnew_0210.api_breaking>` and :ref:`deprecations
 New features
 ~~~~~~~~~~~~
 
+- DataFrame.interpolate() has a new setting: limit_direction='inside'.
+    This will cause the interpolation to fill missing values only when
+    the missing value is surounded by valid values.  It is useful when
```
Comment (crayxt) [inline doc/source/whatsnew/v0.21.0.txt]: surrounded

**Comment by jreback** [comment]
pls add tests for any changes.

**Comment by WBare** [comment]
Sorry for the oversight. I will get the tests added. I also realized I
probably should have done two separate pulls.  On for the fix and one for
the enhancement.

Also, Baurzhan suggested a change from limit_direction='inside' to
surrounded.

Any thoughts on this? I'm thinking 'inside' sounds more like a direction,
but I will make it whatever.

**Comment by WBare** [comment]
I've added required tests for the limit direction feature.  I see them in my fork and I see them referenced in the pull request.

It should (I think) be ready to go, but I'm not sure if I need to do anything to update the pull.  If I've missed anything else, just let me know.  This is my first contribution, so please bear with me.

Thanks

**Comment by WBare** [comment]
Hi Guys,

Looks like I got all checks passed, coverage increased with new tests, I got my spelling error fixed, and I currently have no conflicts with the base branch.

May I get a review on this?  It is actually a fairly simple change.  I just had to learn the process for submitting.

Thanks!

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -19,7 +19,11 @@ Check the :ref:`API Changes <whatsnew_0210.api_breaking>` and :ref:`deprecations
 
 New features
 ~~~~~~~~~~~~
-
+- DataFrame.interpolate() has a new setting: limit_direction='inside'.
+    This will cause the interpolation to fill missing values only when
+    the missing value is surrounded by valid values.  It is useful when 
+    a series needs to be interpolated, but must not expand into NaN
+    values that were outside the range of the original series. (GH16284)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: The reference to the issue should be like

    :issue:`16284`

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -118,6 +122,11 @@ Reshaping
 Numeric
 ^^^^^^^
 
+- DataFrame.interpolate was not respecting limit_direction when
+    limit=0 (unlimited).  Specifically, it would always use
+    limit_direction='forward' even when specified otherwise. Now
+    default limit=0 will work with other directions. (GH16282)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: Same comment about the reference

File: pandas/tests/series/test_missing.py [hunk 4]
```diff
@@ -931,6 +931,40 @@ def test_interp_limit_forward(self):
                                limit_direction='FORWARD')
         assert_series_equal(result, expected)
 
+    def test_interp_limit_inside(self):
+        # these test are for issue #16307 'inside' direction
```
Comment (TomAugspurger) [inline pandas/tests/series/test_missing.py]: You can just change this comment to `GH #16307` or the link to the issue.

**Review by TomAugspurger — COMMENTED** [review]
The new `limit_direction='inside'` seems a bit confusing to me. We use that to refer to the number of consecutive NaNs to fill. With this change, it's either that, or a special case for "out of valid bounds".

How about we add a new keyword that mimics `fill_value` from https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d. We can call it either `fill_value` or `extrapolate`. I think that will require more discussion that the bugfix, so could you split it off of this change and make a new PR?

**Comment by WBare** [comment]
Hi Tom,

Please help me understand your comment.  There are currently two parameters in use:
limit=n  This is the NUMBER of missing values to fill in a given direction.
limit_direction='param' is an already existing parameter indicating the direction in which values should be filled.  Specifically, to the already existing values, I added a new direction of 'inside'.
- **forward** - fill forward including values past the last non-NaN
- **backward** - fill backward including values before the last non-NaN
- **both** - fill both directions including values before and after the last non-NaN
- ADD: _**inside**_ - fill values that are surrounded by non-NaN (will only fill missing values that are surrounded by non-NaN and thus will not expand size of the series).

The new direction 'inside' fits with the other existing directions.  Please help me understand the need to add a new parameter.

Thanks!

**Comment by TomAugspurger** [comment]
Your test kind of shows it (`s.interpolate(method='linear', limit=1, limit_direction='inside'):

```python
s = Series([np.nan, 1, 3, np.nan, np.nan, np.nan, 11, np.nan])
expected = Series([np.nan, 1, 3, 5, np.nan, 9, 11, np.nan])
```

It looks like that's a `limit_direction="both"`, but without extrapolate. You could imagine a case where the user wants to only fill forward (`limit_direction="forward"`) but not extrapolate, so their expected is

```python
expected = Series([np.nan, 1, 3, 5, np.nan, np.nan, 11, np.nan])
#                                           ^ this one changed to still be nan
```

Basically, extrapolation is different that interpolation ;)

**Comment by WBare** [comment]
Ah... right.  Yes, I agree with you.  My key issue was actually with the fact that we are not differentiating between extrapolating and interpolating.  I was fixing one case, but a general solution would cover all the cases.

By default scypi interpolates, and by default pandas interpolates AND extrapolates.

Perhaps:
limit_type=None (default to interpolate and extrapolate as it is today)
limit_type=interpolate (only fills inside values)
limit_type=extrapolate (only fills outside values)

Since I'm in the middle of this, I'd be happy to do this if you guys want it done.  Just let me know what the core contributors decide.

In the mean time, I will revert my code to master and just apply code required for the limit=n bug fix (I should not have combined these in the first place), then submit changes again.

Thanks!

**Comment by TomAugspurger** [comment]
@WBare let's move the extrapolate discussion back to https://github.com/pandas-dev/pandas/issues/16284

**Pull request closed without merging**